target/linux/ipq40xx/base-files/lib/upgrade/zyxel.sh

```diff
@@ -0,0 +1,107 @@
+#
+# Copyright (C) 2016 lede-project.org
+#
+
+zyxel_get_rootfs() {
+	local rootfsdev
+
+	if read cmdline < /proc/cmdline; then
+		case "$cmdline" in
+			*root=*)
+				rootfsdev="${cmdline##*root=}"
+				rootfsdev="${rootfsdev%% *}"
+			;;
+		esac
+
+		echo "${rootfsdev}"
+	fi
+}
+
+zyxel_do_upgrade() {
+	local tar_file="$1"
+	local board=$(board_name)
+	local board_dir=$(tar tf $tar_file | grep -m 1 '^sysupgrade-.*/$')
+	board_dir=${board_dir%/}
+
+	case "$board" in
+	zyxel,wsq50)
+		# Identify our rootfs and kernel partitions
+		local rootfs="$(zyxel_get_rootfs)"
+		local kernel=
+		local loop_supported=
+
+		[ -b "${rootfs}" ] || return 1
+		[ -f "/usr/sbin/losetup" ] && loop_supported=true
+
+		case "$rootfs" in
+			"/dev/mmcblk0p5")
+				kernel="/dev/mmcblk0p4"
```

Comment:
Please no hardcoded partitions

Reply:
This was based off the similar script in https://github.com/openwrt/openwrt/blob/main/target/linux/ipq806x/base-files/lib/upgrade/zyxel.sh with hard-coded partitions - I opted to do this differently to allow rolling back to stock firmware, but it still needs to know where it's installed - this is a specificity of the partition layout of the WSQ50

Reply:
But, they have labels in the partition table, dont they?

So, why not find the partition by the label instead of hardcoding them?

Reply:
Cant you use `find_mmc_part` here?

Reply:
Good shout! Swapping to that now :)

Reply:
This has been fixed - Swapped to `find_mmc_part` like you suggested, still needed to use the initial paths in the case statement though